sqlglot/optimizer/qualify_columns.py

```diff
@@ -416,13 +417,25 @@ def get_table(self, column_name: str) -> t.Optional[exp.Identifier]:
         table_name = self._unambiguous_columns.get(column_name)
 
         if not table_name:
-            sources_without_schema = tuple(
-                source
-                for source, columns in self._get_all_source_columns().items()
-                if not columns or "*" in columns
-            )
-            if len(sources_without_schema) == 1:
-                table_name = sources_without_schema[0]
+            source_name_to_alias = {}
+
+            # Set of pairs (source_name, is_table)
+            distinct_sources_without_schema = set()
```

Comment:
I used a pair here in order to distinguish between `Scope` and `Table` sources, because otherwise we'd get some erroneous transformations in some cases. Consider the example:

```python
SELECT o FROM e AS t1 JOIN (SELECT * FROM t2 AS t2) AS e;
```

Here, the original table name of `t1` is `e`, which is also used as the alias for a `Scope` source, hence we must make sure not to collapse these two different sources into a single one.